libraries/iterate/src/iterate.js

```diff
@@ -76,103 +70,105 @@ class Iterate extends events.EventEmitter {
+    let timeoutRejector = new Promise((resolve, reject) => {
+      watchdog.on('expired', () => {
+        reject(new Error('watchdog exceeded'));
+      });
+
+      maxIterationTimeTimer = setTimeout(() => {
+        reject(new Error('Iteration exceeded maximum time allowed'));
+      }, this.maxIterationTime);
+    });
+
     try {
-      debug('running handler');
-      let start = new Date();
-      let watchdog = new WatchDog(this.watchdogTime);
-      let maxIterationTimeTimer;
-
-      // build a promise that will reject when either the watchdog
-      // times out or the maxIterationTimeTimer expires
-      let timeoutRejector = new Promise((resolve, reject) => {
-        watchdog.on('expired', () => {
-          reject(new Error('watchdog exceeded'));
-        });
+      watchdog.start();
+      await Promise.race([
+        timeoutRejector,
+        Promise.resolve(this.handler(watchdog, this.sharedState)),
+      ]);
+    } finally {
+      // stop the timers regardless of success or failure
+      clearTimeout(maxIterationTimeTimer);
+      watchdog.stop();
+    }
 
-        maxIterationTimeTimer = setTimeout(() => {
-          reject(new Error('Iteration exceeded maximum time allowed'));
-        }, this.maxIterationTime);
-      });
+    let duration = new Date() - start;
+    if (this.minIterationTime > 0 && duration < this.minIterationTime) {
+      throw new Error('Handler duration was less than minIterationTime');
+    }
+  }
 
-      try {
-        watchdog.start();
-        await Promise.race([
-          timeoutRejector,
-          Promise.resolve(this.handler(watchdog, this.sharedState)),
-        ]);
-      } finally {
-        // stop the timers regardless of success or failure
-        clearTimeout(maxIterationTimeTimer);
-        watchdog.stop();
-      }
+  // run a single iteration, throwing any errors
+  async iterate() {
+    let currentIteration = 0;
+    let failures = [];
+    while (true) {
+      currentIteration++;
+      let iterError;
 
-      // TODO: do this timing the better way
-      let diff = new Date() - start;
+      this.emit('iteration-start');
 
-      // Let's check that if we have a minimum threshold for handler activity
-      // time, and mark as failure when we exceed it
-      if (this.minIterationTime > 0 && diff < this.minIterationTime) {
-        throw new Error('Minimum threshold for handler execution not met');
+      const start = process.hrtime();
+      try {
+        await this.single_iteration();
+      } catch (err) {
+        iterError = err;
       }
+      const d = process.hrtime(start);
+      const duration = d[0] * 1000 + d[1] / 1000000;
 
-      // Wait until we've done all the checks to emit success
-      this.emit('iteration-success');
-      debug(`ran handler in ${diff} seconds`);
+      this.emit(iterError ? 'iteration-failure' : 'iteration-success');
       if (this.monitor) {
         this.monitor.info('iteration', {
```

Comment:
Should we make this a registered log type? Not blocking landing this.

Reply:
Yeah, I think so.  I figured that would find itself in your dragnet..